ServerCore/Helpers/NotificationHelper.cs

```diff
@@ -0,0 +1,131 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using ServerCore.ServerMessages;
+
+/// <summary>
+/// EventArgs for callers that register for events.
+/// </summary>
+public class NotificationEventArgs : EventArgs
+{
+    /// <summary>
+    /// The notification data
+    /// </summary>
+    public Notification Notification { get; set; }
+}
+
+/// <summary>
+/// Helper object for subscribing to notifications and sending notifications.
+/// </summary>
+public class NotificationHelper
+{
+    private ConcurrentDictionary<int, ConcurrentDictionary<NotificationListener, byte>> listenersByEvent = new ConcurrentDictionary<int, ConcurrentDictionary<NotificationListener, byte>>();
+    private ConcurrentDictionary<int, ConcurrentDictionary<NotificationListener, byte>> listenersByTeam = new ConcurrentDictionary<int, ConcurrentDictionary<NotificationListener, byte>>();
+    private ConcurrentDictionary<int, ConcurrentDictionary<NotificationListener, byte>> listenersByPlayer = new ConcurrentDictionary<int, ConcurrentDictionary<NotificationListener, byte>>();
+
+    public NotificationHelper(ServerMessageListener messageListener)
+    {
+        messageListener.OnNotification += ReceiveNotification;
+    }
+
+    /// <summary>
+    /// Notification receiver, called by the sender (locally if isDev, via SignalR if deployed)
+    /// </summary>
+    /// <param name="notification">The notification received</param>
+    private Task ReceiveNotification(Notification notification)
```

Comment:
I'm a little confused about this returning a task that never gets used. Should this either be sync or flow async down to NotificationRecieved?

Reply:
Perhaps, but I can't figure out how to do it. What does an async event handler look like?

Reply:
Here's an example from Presence https://github.com/PuzzleServer/mainpuzzleserver/blob/ffed1c36e066878c5cf0193fbf56553b69a3faff/ServerCore/ServerMessages/PresenceStore.cs#L65

Reply:
So no sender, no EventHandler or EventArgs, and it still works?

I also assume the true magic is in this line:

`public event Func<IDictionary<Guid, PresenceModel>, Task> OnTeamPuzzlePresenceChange;`

Reply:
Correct. You can use events and delegates with any signature you like. EventHandler<T> is a predefined signature you can use, as is Func<T> and Action<T> (The difference between Func and Action is Action returns void, Func returns the last type you list)

Reply:
Works, thanks!